src/security/KeyData.cc

```diff
@@ -78,13 +79,51 @@ Security::KeyData::loadX509CertFromFile()
     return bool(cert);
 }
 
+#if USE_OPENSSL || USE_GNUTLS
+static void
+verifyChainCert(Security::CertPointer &ca, Security::CertPointer &latestCert, Security::CertList &chain)
+{
+    auto nameStr = Security::CertSubjectName(ca);
+    Security::ErrorCode checkCode;
+#if !USE_OPENSSL || TLS_CHAIN_NO_SELFSIGNED
+    // self-signed certificates are not valid in a sent chain
+    if (Security::CertIssuerCheck(ca, ca, checkCode)) {
+        debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
+        return;
+    }
+#endif
+
+    // checks that the chained certs are actually part of a chain for validating cert
+    if (Security::CertIssuerCheck(latestCert, ca, checkCode)) {
+        debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
+        // OpenSSL API requires that we order certificates such that the
+        // chain can be appended directly into the on-wire traffic.
+        latestCert = ca;
+        chain.emplace_front(latestCert);
+    } else {
+        debugs(83, DBG_PARSE_NOTE(2), "Ignoring non-issuer CA " << nameStr << ": " << Security::VerifyErrorString(checkCode) << " (" << checkCode << ")");
+    }
+}
+#endif
+
 /**
  * Read certificate from file.
  * See also: Ssl::ReadX509Certificate function, gadgets.cc file
  */
 void
 Security::KeyData::loadX509ChainFromFile()
 {
+    ErrorCode checkCode;
+    if (CertIssuerCheck(cert, cert, checkCode)) {
```

Comment:
Done, but see above about throwing.

Reply:
for specific property/state checks like this one, we should use adjective-based or be-adjective-based names like SelfSignedCert(), CertIsSelfSigned(), or similar as opposed to noun-based CertSelfSignedCheck(). We want to be able to write readable/intuitive `if (the certificate is self signed)` rather than misleading/confusing `if (we can check whether the certificate is self signed)` statements.